File: 079-word-search/word-search.py
# -*- coding:utf-8 -*-


#
# Given a 2D board and a word, find if the word exists in the grid.
#
#
# The word can be constructed from letters of sequentially adjacent cell, where "adjacent" cells are those horizontally or vertically neighboring. The same letter cell may not be used more than once.
#
#
#
# For example,
# Given board = 
#
# [
#   [&#39;A&#39;,&#39;B&#39;,&#39;C&#39;,&#39;E&#39;],
#   [&#39;S&#39;,&#39;F&#39;,&#39;C&#39;,&#39;S&#39;],
#   [&#39;A&#39;,&#39;D&#39;,&#39;E&#39;,&#39;E&#39;]
# ]
#
#
# word = "ABCCED", -&gt; returns true,
# word = "SEE", -&gt; returns true,
# word = "ABCB", -&gt; returns false.


class Solution(object):
    def exist(self, board, word):
        """
        :type board: List[List[str]]
        :type word: str
        :rtype: bool
        """
        if not board:
            return False
        for i in xrange(len(board)):
            for j in xrange(len(board[0])):
                if self.dfs(board, i, j, word):
                    return True
        return False
    
    # check whether can find word, start at (i,j) position    
    def dfs(self, board, i, j, word):
        if len(word) == 0: # all the characters are checked
            return True
        if i<0 or i>=len(board) or j<0 or j>=len(board[0]) or word[0]!=board[i][j]:
            return False
        tmp = board[i][j]  # first character is found, check the remaining part
        board[i][j] = "#"  # avoid visit agian 
        # check whether can find "word" along one direction
        res = self.dfs(board, i+1, j, word[1:]) or self.dfs(board, i-1, j, word[1:]) \
        or self.dfs(board, i, j+1, word[1:]) or self.dfs(board, i, j-1, word[1:])
        board[i][j] = tmp
        return res
                
        
        
        

File: 079-word-search/word-search.cpp
// Given a 2D board and a word, find if the word exists in the grid.
//
// The word can be constructed from letters of sequentially adjacent cell, where "adjacent" cells are those horizontally or vertically neighboring. The same letter cell may not be used more than once.
//
// Example:
//
//
// board =
// [
//   ['A','B','C','E'],
//   ['S','F','C','S'],
//   ['A','D','E','E']
// ]
//
// Given word = "ABCCED", return true.
// Given word = "SEE", return true.
// Given word = "ABCB", return false.
//
//


class Solution {
public:
    bool exist(vector<vector<char>>& board, string word) {
        int m = board.size();
        if (!m) return false;
        int n = board[0].size();
        if (!n) return false;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (dfs(board, word, i, j, 0)) {
                    return true;
                }
            }
        }
        return false;
    }
    vector<int> dir{0, 1, 0, -1, 0};
    bool dfs(vector<vector<char>>& board, string word, int i, int j, int idx) {
        if (i < 0 || j < 0 || i >= board.size() || j >= board[0].size() || idx >= word.size() || board[i][j] != word[idx]) {
            return false;
        }
        if (idx == word.size() - 1) {
            return true;
        }
        char c = board[i][j];
        board[i][j] = '#';
        for (int k = 0; k < 4; k++) {
            if (dfs(board, word, i + dir[k], j + dir[k + 1], idx + 1)) {
                board[i][j] = c;
                return true;
            }
        }
        board[i][j] = c;
        return false;
    }
};
